dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/managedobject/CDIEJBManagedObjectFactoryImpl.java

```diff
@@ -89,28 +91,99 @@ public ManagedObjectContext createContext() {
+                    Iterator<WebSphereBeanDeploymentArchive> itr = children.iterator();
+                    while (ejbDescriptor == null && itr.hasNext()) {
+                        bda = itr.next();
+                        beanManager = (WeldManager) bda.getBeanManager();
+                        ejbDescriptor = beanManager.getEjbDescriptor(this.ejbName);
+                        bean = beanManager.getBean(ejbDescriptor);
+                        if (! bean.getBeanClass().equals(getManagedObjectClass())) {
+                            partialMatchDescriptors.add(ejbDescriptor);
+                            ejbDescriptor = null;
+                            bean = null;
+                        }
+                    }
+
+                    // check all accessible BDAs if we haven't found the ejbDescriptor in descendant BDAs
+                    if (ejbDescriptor == null) {
+                        children = bda.getWebSphereBeanDeploymentArchives();
+                        itr = children.iterator();
+                        while (ejbDescriptor == null && itr.hasNext()) {
+                            bda = itr.next();
+                            beanManager = (WeldManager) bda.getBeanManager();
+                            ejbDescriptor = beanManager.getEjbDescriptor(this.ejbName);
+                            bean = beanManager.getBean(ejbDescriptor);
+                            if (! bean.getBeanClass().equals(getManagedObjectClass())) {
+                                partialMatchDescriptors.add(ejbDescriptor);
+                                ejbDescriptor = null;
+                                bean = null;
+                            }
+                        }
                     }
                 }
+
+                if (ejbDescriptor != null) {
+                    this.ejbDescriptor = ejbDescriptor;
+                    this.ejbBeanManager = beanManager;
+                    this.ejbBDA = bda;
+                    this.ejbBean = bean;
+                }
+
+            }
+
+            if (this.ejbDescriptor == null && partialMatchDescriptors.isEmpty()) {
+                Tr.error(tc, "Could not find an EjbDescriptor for : " + this.ejbName);
+            } else if (this.ejbDescriptor == null && ! partialMatchDescriptors.isEmpty()) {
+                String listStr = "[";
+                for (EjbDescriptor<T> desc : partialMatchDescriptors) {
+                    listStr += desc.getEjbName() + ",";
+                }
+                listStr = listStr.substring(0, listStr.length() - 1);
+                listStr += "]";
+                Tr.error(tc, "Found the following EjbDescriptors but they produced the wrong bean " + listStr);
             }
 
-            this.ejbBeanLookupComplete = true;
+            ejbBeanLookupComplete = true;
 
             if (this.ejbBean != null) {
                 if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
```

Comment:
Should this check not happen before we set `ejbBeanLookupComplete = true`? (and before we initialize the rest of the fields)

Reply:
Good question. In the old version three of the four fields could be set from getEJBDescriptor without ever calling getBean

Do we want to preserve that old behaviour? 

But then we could set this.ejbBean to a bean, throw an exception because `this.ejbBean.getBeanClass().equals(getManagedObjectClass()` then the next time you call get bean ejbBeanLookupComplete is true so you get the bean with no exception which is _horrific_

So I think we do it the sane way, at least until we run some builds and see if anything breaks.